internal/espressoreader/espresso_reader.go

```diff
@@ -196,158 +209,228 @@ func (e *EspressoReader) readL1(ctx context.Context, app evmreader.TypeExportApp
+		return nil, err
+	}
+	appAddressStr, ok := typedData.Message["app"].(string)
+	if !ok {
+		slog.Error("failed to cast app address to string")
+		return nil, err
+	}
+
+	slog.Info("Espresso input", "msgSender", msgSender, "nonce", nonce, "payload", payload, "appAddrss", appAddressStr, "tx-id", sigHash)
+
+	payloadBytes := []byte(payload)
+	if strings.HasPrefix(payload, "0x") {
+		payload = payload[2:] // remove 0x
+		payloadBytes, err = hex.DecodeString(payload)
+		if err != nil {
+			slog.Error("failed to decode hex string", "error", err)
+			return nil, err
+		}
+	}
+	espressoInput := EspressoInput{
+		app:       strings.ToLower(appAddressStr),
+		data:      payloadBytes,
+		nonce:     nonce,
+		msgSender: msgSender,
+		sigHash:   sigHash,
+	}
+
+	return &espressoInput, nil
+}
+
+func (e *EspressoReader) encodeEvmAdvance(espressoInput *EspressoInput, chainId *big.Int, app common.Address,
+	indexUint64 uint64, l1FinalizedLatestHeightBig *big.Int, l1FinalizedTimestampBig *big.Int,
+	prevRandao *big.Int,
+) ([]byte, error) {
+	payloadBytes := espressoInput.data
+	index := &big.Int{}
+	index.SetUint64(indexUint64)
+	payloadAbi, err := e.evmReader.IOAbi.Pack("EvmAdvance", chainId, app, espressoInput.msgSender, l1FinalizedLatestHeightBig, l1FinalizedTimestampBig, prevRandao, index, payloadBytes)
+	if err != nil {
+		slog.Error("failed to abi encode", "error", err)
+		return nil, err
+	}
+	return payloadAbi, nil
+}
+
+func toBigInt(value uint64) *big.Int {
+	bigInt := &big.Int{}
+	bigInt.SetUint64(value)
+	return bigInt
+}
+
+func (e *EspressoReader) buildInput(ctx context.Context,
+	espressoInput *EspressoInput, chainId *big.Int,
+	app common.Address,
+	l1FinalizedLatestHeightBig *big.Int,
+	l1FinalizedTimestampBig *big.Int,
+	prevRandao *big.Int,
+	l1FinalizedLatestHeight uint64,
+) (*model.Input, error) {
+	indexUint64, err := e.repository.GetInputIndex(ctx, espressoInput.app)
+	if err != nil {
+		slog.Error("failed to read index", "app", espressoInput.app, "error", err)
+	}
+
+	// abi encode payload
+	payloadAbi, err := e.encodeEvmAdvance(espressoInput, chainId, app, indexUint64, l1FinalizedLatestHeightBig, l1FinalizedTimestampBig, prevRandao)
+	if err != nil {
+		slog.Error("failed to abi encode", "error", err)
+		return nil, err
+	}
+	// build input
+	sigHashHexBytes, err := hex.DecodeString(espressoInput.sigHash[2:])
+	if err != nil {
+		slog.Error("could not obtain bytes for tx-id", "err", err)
+		return nil, err
+	}
+	input := model.Input{
+		Index:                indexUint64,
+		Status:               model.InputCompletionStatus_None,
+		RawData:              payloadAbi,
+		BlockNumber:          l1FinalizedLatestHeight,
+		TransactionReference: common.BytesToHash(sigHashHexBytes),
+	}
+	return &input, nil
+}
+
+func (e *EspressoReader) isNonceValid(ctx context.Context, espressoInput *EspressoInput) (bool, error) {
+	nonceInDb, err := e.repository.GetEspressoNonce(ctx, espressoInput.msgSender.Hex(), espressoInput.app)
+	if err != nil {
```

Comment:
Here, I maintain the same behavior, but it's better to handle the error by either retrying or halting.